File: Ningin/src/sceneSystem/components/transform.cpp
#include "Transform.h"

Transform::Transform(Vector3& position, Vector3& scale, Vector3& rotation) : position(position),
	scale(scale), rotation(rotation) {}

Transform::Transform() : position(0.0f, 0.0f, 0.0f), scale(1.0f, 1.0f, 1.0f), rotation(0.0f, 0.0f, 0.0f) {}

void TransformSetPosition(Transform& transform, Vector3& position)
{
	transform.position = position;
}

void TransformSetRotation(Transform& transform, Vector3& rotation)
{
	transform.rotation = rotation;
}

void TransformSetScale(Transform& transform, Vector3& scale)
{
	transform.scale = scale;
}

void TransformGetPosition(Transform& transform, Vector3& out)
{
	out = transform.position;
}

void TransformGetRotation(Transform& transform, Vector3& out)
{
	out = transform.rotation;
}

void TransformGetScale(Transform& transform, Vector3& out) {
	out = transform.scale;
}

void TransformToString(Transform& transform, string& out)
{
	out = format("Transform - Position: x: {}, y: {}, z: {}\nRotation: x: {}, y: {}, z: {}\nScale: x: {}, y: {}, z: {}",
		transform.position.x, transform.position.y, transform.position.z, transform.rotation.x,
		transform.rotation.y, transform.rotation.z, transform.scale.x, transform.scale.y, transform.scale.z);
}

File: NinginCore/components/Transform.cs
using System;

namespace NinginCore
{
    public class Transform : Component
    {

        public Vector3 Position
        {
            get
            {
                InternalCalls.TransformGetPosition(this.data, out Vector3 result);
                return result;
            }

            set
            {
                InternalCalls.TransformSetPosition(this.data, ref value);
            }
        }

        public Vector3 Rotation
        {
            get
            {
                InternalCalls.TransformGetRotation(this.data, out Vector3 result);
                return result;
            }

            set
            {
                InternalCalls.TransformSetRotation(this.data, ref value);
            }
        }

        public Vector3 Scale
        {
            get
            {
                InternalCalls.TransformGetScale(this.data, out Vector3 result);
                return result;
            }

            set
            {
                InternalCalls.TransformSetScale(this.data, ref value);
            }
        }

        public override string ToString()
        {
            InternalCalls.MonoTransformToString(this.data, out string result);
            return result;
        }
    }
}
